### PR TITLE
feat(linter): support useGuardForIn

### PR DESCRIPTION
**Pull request by fireairforce**
## Summary

closes: #3933

### CODE REVIEW

**Comment by fireairforce** [comment]
sorry i miss the rule name convention for biome, i update the lint name now, please review again when you free~ @ematipico

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    context::RuleContext, declare_lint_rule, Ast, Rule, RuleDiagnostic, RuleSource,
+};
+use biome_console::markup;
+use biome_js_syntax::{AnyJsStatement, JsForInStatement};
+use biome_rowan::{AstNode, AstNodeList};
+
+declare_lint_rule! {
+    ///
+    /// Require `for-in` loops to include an `if` statement
+    ///
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs:11]: Let's add more documentation, please. Let's explain why an `if` statement is required by this rule.

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+for (var x in o) { if (x) f(); g(); }
+for (var x in o) { foo() }
+for (var x in o) foo();
+```
+
+# Diagnostics
+```
+invalid.js:1:1 lint/nursery/useGuardForIn ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+
+  ! Require `for-in` loops to include an `if` statement.
```
Comment (ematipico) [inline crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap:19]: The message should be improved. In this case, there's a `if` statement, so I was a user, I wouldn't know what to do.  I would suggest an alternative, but at the moment the documentation is poor, so I don't understand why there's a diagnostic for this case.

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  
+  > 1 │ for (var x in o) { if (x) { f(); continue; } g(); }
+      │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    2 │ for (var x in o) { if (x) { continue; f(); } g(); }
+    3 │ for (var x in o) { if (x) { f(); } g(); }
+  
+  i The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype.
```
Comment (ematipico) [inline crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap:26]: Our linter diagnostics must follow the rule pillars: https://biomejs.dev/linter/#rule-pillars 

This is also present in our contribution guide, please get familiar with it.

In this case, we fail to explain the error, the second pillar.

**Review by ematipico — CHANGES_REQUESTED** [review]
The code might be good, but unfortunately the documentation is nearly absent, and the diagnostics should provide more information to the user.

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs (hunk 1 above)
Comment (fireairforce) [reply crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs:11]: ok

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap (hunk 2 above)
Comment (fireairforce) [reply crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap:19]: sorry for that, i will add more detailed message for it.

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs (hunk 1 above)
Comment (fireairforce) [reply crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs]: Add a more detailed desc here

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap (hunk 2 above)
Comment (fireairforce) [reply crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap]: add a message here, please review again for you free.

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap (hunk 3 above)
Comment (fireairforce) [reply crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/invalid.js.snap]: Sorry, I referred too much to eslint here.. its message is very simple. i updated the message here..

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+};
+use biome_console::markup;
+use biome_js_syntax::{AnyJsStatement, JsForInStatement};
+use biome_rowan::{AstNode, AstNodeList};
+
+declare_lint_rule! {
+    ///
+    /// Require `for-in` loops to include an `if` statement.
+    ///
+    /// Looping over objects with a for in loop will include properties that are inherited through the prototype chain.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs:12]:
```suggestion
    /// Looping over objects with a `for-in` loop will include properties inherited through the prototype chain.
```

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    /// This behavior can lead to unexpected items in your for loop.
+    ///
+    /// For codebases that do not support ES2022, `Object.prototype.hasOwnProperty.call(foo, key)` can be used as a check that the property is not inherited.
+    ///
+    /// For codebases that do support ES2022, `Object.hasOwn(foo, key)` can be used as a shorter alternative.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs:17]:
```suggestion
    /// For codebases that do support ES2022, `Object.hasOwn(foo, key)` can be used as a shorter and more reliable alternative.
```

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                rule_category!(),
+                node.range(),
+                markup! {
+                    "The body of a for-in should be wrapped in an `if` statement."
+                },
+            )
+            .note(markup! {
+                "Looping over the object with for-in loop  will include properties that are inherited through the prototype chain, the behaviour can lead to some unexpected items in your loop."
+            }).note(markup! {
+                "To resolve this issue, add an if statement like `if (Object.hasOwn(foo, key)) {...}` to wrapper the body of for-in loop. "
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs:133]:
```suggestion
                "To resolve this issue, add an if statement like `if (Object.hasOwn(foo, key)) {...}` to filter out the extraneous properties. "
```

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/valid.js.snap
Comment (ematipico) [inline crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/valid.js.snap:1]: Shouldn't we add cases where we have a `for..in` loop with `Object.hasOwn(foo, key)`, for example?

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs (hunk 4 above)
Comment (fireairforce) [reply crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs]: done

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs (hunk 5 above)
Comment (fireairforce) [reply crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs]: done

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs (hunk 6 above)
Comment (fireairforce) [reply crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs]: done~

File: crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/valid.js.snap
Comment (fireairforce) [reply crates/biome_js_analyze/tests/specs/nursery/useGuardForIn/valid.js.snap:1]: updated~ thanks for review

File: crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, Ast, Rule, RuleDiagnostic, RuleSource,
+};
+use biome_console::markup;
+use biome_js_syntax::{AnyJsStatement, JsForInStatement};
+use biome_rowan::{AstNode, AstNodeList};
+
+declare_lint_rule! {
+    ///
+    /// Require `for-in` loops to include an `if` statement.
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/nursery/use_guard_for_in.rs:10]:
```suggestion
    /// Require `for-in` loops to include an `if` statement.
```

**Review by ematipico — APPROVED** [review]

**Comment by fireairforce** [comment]
i solve the conflict at my local, maybe the mr can be merged later?

**Pull request merged**